cli/src/pcluster/models/cluster.py

```diff
@@ -12,939 +12,339 @@
+                    )
+                )
+                # config version will be stored in DB by the cookbook at the first update
+                self.config_version = result.get("VersionId")
 
-    @property
-    def vpc_id(self):
-        """Return the VPC of the cluster."""
-        return (
-            AWSApi.instance()
-            .ec2.describe_subnets(SubnetIds=[self.head_node.networking.subnet_id])
-            .get("Subnets")[0]
-            .get("VpcId")
-        )
+        except Exception as e:
+            raise ClusterActionError(
+                f"Unable to upload cluster resources to the S3 bucket {self.bucket.name} due to exception: {e}"
+            )
 
     @property
-    def ami_id(self):
-        """Get the image id of the cluster."""
-        return (
-            self.image.custom_ami
-            if self.image.custom_ami
-            else AWSApi.instance().ec2.get_official_image_id(self.image.os, self.head_node.architecture)
-        )
+    def template_url(self):
+        """Return template url."""
+        if self.config.dev_settings and self.config.dev_settings.cluster_template:
+            # template provided by the user
+            template_url = self.config.dev_settings.cluster_template
+        else:
+            # default template
+            template_url = "https://{bucket_name}.s3.{region}.amazonaws.com{partition_suffix}/{template_key}".format(
+                bucket_name=self.bucket.name,
+                region=self.config.region,
+                partition_suffix=".cn" if self.config.region.startswith("cn") else "",
+                template_key=self._get_default_template_key(),
+            )
+        return template_url
+
+    def _get_default_template_key(self):
+        return f"{self.bucket.artifact_directory}/templates/aws-parallelcluster.cfn.yaml"
+
+    def _get_config_key(self):
+        return f"{self.bucket.artifact_directory}/configs/cluster-config.json"
+
+    def delete(self, keep_logs: bool = True):
+        """Delete cluster."""
+        try:
+            self.stack.delete(keep_logs)
+        except Exception as e:
+            self._terminate_nodes()
+            raise ClusterActionError(f"Cluster deletion failed. Error: {e}")
+
+    def _terminate_nodes(self):
+        try:
+            LOGGER.info("\nChecking if there are running compute nodes that require termination...")
+            filters = [
+                {"Name": "tag:Application", "Values": [self.stack_name]},
+                {"Name": "instance-state-name", "Values": ["pending", "running", "stopping", "stopped"]},
+                {"Name": "tag:aws-parallelcluster-node-type", "Values": [str(NodeType.compute)]},
+            ]
+            result = AWSApi().instance().ec2.describe_instances(filters)
+            instances = []
+            for instance in result.get("Instances", []):
+                instances.append(instance.get("InstanceId"))
+
+            for instance_ids in grouper(instances, 100):
```

Comment:
Why using grouper?

Reply:
Because AWSApi automatically perform a pagination and return the full list.
Here we need to group the instances in small_group because it's suggested by boto3 doc: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.terminate_instances